README.md
**DeCiDa Data Analysis and Procedural Simulation Python Library**
=================================================================

DeCiDa == Device and Circuit Data Analysis.

`DeCiDa man pages <http://www.decida.org/PyDeCiDa_man_pages>`_

DeCiDa is a Python Library of functions and classes for
device characterization, circuit design and data analysis.

DeCida is written in pure Python (2.7, and 3), and requires no code-compilation.
It is portable to any operating system where Python is installed, and runs
under MacOS, Windows, Cygwin, and Linux.  Many DeCiDa classes and functions
require the numpy and Tkinter packages.

This version of DeCiDa provides support for Python 3 and can be run under
a Python 3 environment, as well as a Python 2.7 environment.
The following packages are now required: future, six, numpy and matplotlib.
Please let me know if there are any issues that you find.

DeCiDa uses matplotlib XY-plot rendering by default.
To use the former XY-plot rendering, add the option -xmat
to dataview, plotter, pllss, and pll_phase_noise.

For circuit analysis, DeCiDa provides a flexible scripting class for
performing simulations with various circuit simulation tools, such as
(Synopsys) HSpice, (Cadence) Spectre, (Silvaco) SmartSpice,
(UC Berkeley) NGspice and others.
The Tckt class is used to provide a database of process corners for
each project, and to provide netlist templating for performing procedural
simulations and post-processing.
Scripts using Tckt access the database to obtain
the corner conditions, modify the netlist, and allow full Python looping
structures for running the simulation, viewing and analyzing
the simulated data.

For data analysis, DeCiDa provides a Data class for reading-in and 
analyzing data in a number of formats, including nutmeg
(NGspice, Spectre, LTspice), CSDF (HSpice), CSV (comma-separated value),
SSV (space-separated value), and others.
The Data object can be viewed using the XYplotm, Histogramx or DataViewm
classes.
DataViewm has commands via menu entries to manipulate and plot the data in
different ways, including frequency versus time, eye- and scope-diagrams
and column calculations.

DeCiDa started out as a Tcl/Tk application for analyzing measurements
of electron devices for performing routine compact-model parameter extraction.
To do this fitting, a least-squares optimization algorithm was used.
This Python version of DeCiDa has a function LevMar (for Levenberg-Marquardt),
based on the mpfit package.  It is still under development.

What is in the distribution
---------------------------

* The decida Python library of functions and classes (./decida).
  This is installed into the site-packages directory.

* A test library under decida.test for testing the distribution (./decida/test).
  This is also installed into the site-packages directory.

* Applications in the distribution bin directory (./bin).
  These are installed into the Python bin directory.

* Tool-specific scripts (./etc):

   * simulation tool wrappers (./etc/wrappers)

      Circuit simulation tool wrappers that DeCiDa interfaces with.
      These are installed into ~/.DeCiDa/bin

   * HTML documentation of the functions and classes (./doc/html).
      This is installed into ~/.DeCiDa/doc

   * cython (./etc/cython)

      Setup scripts for using cython to compile the Data and XYplotm classes.
      These are installed into ~/.DeCiDa/cython

   * dot files (./etc/dot)

      Several resource files to be placed in user home directory
      for Cadence and Python.
      These are installed into ~/.DeCiDa/dot

   * user local lib directory (./etc/lib)

      A place to put user Python code.
      This is set up as ~/.DeCiDa/lib

   * models (./etc/models)

      Case-corners and models for two example technologies from
      the Predictive Technology Models web-site.
      These are installed into ~/.DeCiDa/models

   * projects (./etc/projects)

      Two example project simulation directories (bird and trane).
      These are installed into ~/.DeCiDa/projects

   * Cadence skill files (./etc/skill)

      Several scripts for automatically generating DeCiDa Python
      procedural simulation scripts,
      and verilog test-bench environments.
      These are installed into ~/.DeCiDa/skill

   * stdcell (./etc/stdcell)

      Two example standard cell libraries for the two example
      PTM technologies, from the NangateOpenCell Library.
      These are installed into ~/.DeCiDa/stdcell

   * verilog (./etc/verilog)

      Files for running Cadence NCsim and viewing the results using SimVision.
      These are installed into ~/.DeCiDa/verilog

   * matlab (./etc/matlab)

      Matlab file to implement a Data object with a demo.
      These are installed into ~/.DeCiDa/matlab


DeCiDa applications
-------------------

All of these should be in the path after installation:

+----------------------+---------------------------------------------------------------------------------------------+
| application:         | description:                                                                                |
+======================+=============================================================================================+
| calc                 | scientific calculator                                                                       |
+----------------------+---------------------------------------------------------------------------------------------+
| dataview             | read, plot and analyze data                                                                 |
+----------------------+---------------------------------------------------------------------------------------------+
| plotter              | plot Cartesian, Parametric, or Polar functions                                              |
+----------------------+---------------------------------------------------------------------------------------------+
| twin                 | text editor, with additional capability                                                     |
+----------------------+---------------------------------------------------------------------------------------------+
| gifimg               | create embedded GIF image Python class from a GIF image                                     |
+----------------------+---------------------------------------------------------------------------------------------+
| pllss                | plot PLL small-signal transfer functions, S-domain and Z-domain                             |
+----------------------+---------------------------------------------------------------------------------------------+
| pll_phase_noise      | plot PLL phase noise components and total phase noise                                       |
+----------------------+---------------------------------------------------------------------------------------------+
| ngsp                 | UC Berkeley NGspice gui                                                                     |
+----------------------+---------------------------------------------------------------------------------------------+
| op                   | read Cadence Spectre operating-point analysis, display node voltages and operating points   |
+----------------------+---------------------------------------------------------------------------------------------+
| simvision_csv2col    | convert exported Cadence SimVision csv data to column data                                  |
+----------------------+---------------------------------------------------------------------------------------------+

Simulation tool wrapper scripts
-------------------------------

These scripts are installed in ~/.DeCiDa/bin

+--------------------+--------------------------------------+
| wrapper script:    | description:                         |
+====================+======================================+
| run_hspice         | wrapper to run Synopsys HSpice       |
+--------------------+--------------------------------------+
| run_ngspice        | wrapper to run UC Berkeley NGspice   |
+--------------------+--------------------------------------+
| run_sspice         | wrapper to run Silvaco SmartSpice    |
+--------------------+--------------------------------------+
| run_spectre        | wrapper to run Cadence Spectre       |
+--------------------+--------------------------------------+

Thanks to
---------

1. `Python distribution documentation <http://docs.python.org/2/distutils/index.html>`_ .

2. `Canopy installation documentation <https://support.enthought.com/entries/23389761-Installing-packages-into-Canopy-User-Python-from-the-command-line>`_ .

3. decida/ItclObjectx:

   Concepts from [incr Tcl], described in chapter 2, "Object-Oriented
   Programming with [incr Tcl]," by Michael McLennan, of "Tcl/Tk Tools,"
   Mark Harrision, 1997, O'Reilly. 

4. decida/FrameNotebook and decida/Balloonhelp:

   Adapted from the Tcl/Tk examples in
   Mark Harrison and Michael McLennan, "Effective Tcl/Tk Programming",
   1997, Addison-Wesley.

5. decida/Data.read_nutmeg method:

   Modified from the 
   `read_spice module <http://www.h-renrew.de/h/python_spice/spicedata.html>`_
   from Werner Hoch (python_spice-0.0.3).
    
6. decida/LevMar:

   Modified from the
   `mpfit module <https://code.google.com/p/astrolibpy>`_
   from Sergey Koposov, Craig Markwardt and Mark Rivers (mpfit_2013).

7. bin/gifimg:

   Modified from the 
   `img2pytk module <http://www.3dartist.com/WP/python/code/img2pytk.py>`_
   from Bill Allen (imageEmbedder-1.0).

8. Example model files:

   From `Predictive Technology Model <http://ptm.asu.edu>`_ from the
   Nanoscale Integration and Modeling (NIMO) group at Arizona State University.

9. Example standard cell libraries:

   From `Si2 <https://www.si2.org>`_ openEDA project, Nangate 45nm
   Open Cell Library, a generic open-source, non-manufacturable
   standard-cell library.

10. George Howlett, Michael McLennan, Sani Nassif, Mike Toth and others
    for developing many of the original concepts which are incorporated in
    DeCiDa.

11. Dean Gonzales, Sanquan Song and Phillip Johnson for supplying test
    data files and test-driving DeCiDa.

12. `MatPlotLib <https://matplotlib.org>`_ matplotlib.

13. Barry J. Muldrey, for testing and helping to port to Python 3.

**Installing DeCiDa**
=====================

If you have pip
---------------

* issue this command::

    pip install DeCiDa

Note that the scripts that should be installed in
the python bin directory (dataview, plotter, ...) may not arrive there.
And the home directory directories may not get set up properly.  If this
happens, simply download the distribution and copy these from the untarred
folders.

Otherwise:
----------

Download and prepare the distribution
-------------------------------------

* unzip/untar the distribution::

    tar xvfz DeCiDa.1.1.3-tar.gz

* cd into the distribution directory::

    cd DeCiDa-1.1.3

* you may want to install the DeCiDa html documentation (in ./doc/html)
  to an appropriate place for future reference.
  use a browser to read the documentation, using the url file:// specification
  to point to the index.html file in the html directory.

* manually modify the wrapper scripts in the distribution
  ./etc/wrapper directory (run_*), to point to correct tool locations.
  
  The wrappers have the following references to other tools.  
  Adjust these as needed, as required by your local environment.

+----------------+-------------------------------+
| wrapper:       | expected tool location:       |
+================+===============================+
| run_hspice     | /tools/hspice/bin/hspice      |
+----------------+-------------------------------+
| run_ngspice    | /opt/local/bin/ngspice        |
+----------------+-------------------------------+
| run_sspice     | /tools/silvaco/bin/sspice     |
+----------------+-------------------------------+
| run_spectre    | /tools/cds/bin/spectre        |
+----------------+-------------------------------+

Installing under Anaconda
-------------------------

    refer to: `Managing packages <https://conda.io/docs/user-guide/tasks/manage-pkgs.html>`_

* be sure that Anaconda python is in your path::

    python
    >>> import sys; sys.prefix

  you should see a path like the following::

    /Users/<user>/anaconda (MacOs)

    /home/<user>/anaconda (Linux)

* install the distribution::

    conda install DeCiDa

Installing under Enthought Canopy
---------------------------------

* install in the Canopy Python User Virtual Environment


CHANGES.md
==========
Change Log
==========

v1.0.0, 11-10-2013
------------------

- Initial release.

v1.0.1,  4-20-2014
------------------

- changes to DataViewx.

- changes to Fitter.

v1.0.2, 12-14-2014
------------------

- some changes to skill code

- added new basic modules:

  o pattern_alignment: compare two bit-patterns

  o syscall: simplified system call

- added new classes:
  
  o DialogBase: dialog base class with fixed font

  o MessageDialog: message dialog for analysis reports

  o FormulaCalculator: formula gui calculator class

- Calc:

  o fix binding problem

- Data:

  o read_hspice : read binary, other fixes
 
  o read_nutmeg : partially written data-file read-in

  o added new methods: delays, skews, period_time_average

  o modified jitter analysis

- DataViewx:

  o add analyses: INL/DNL, period-time-average

  o added MessageDialog to DataViewx analysis reports

- PlotBase:

  o add new windowing bindings

  o other minor fixes

- Tckt:

  o architectural changes

  o fixed several issues

- interpolate:

  o fix dictionary issue

- range_sample:

  o fix random issue

v1.0.3,  4-14-2015
------------------

- added new classes:

  o StatusDialog

  o Balloonhelp

- Data.py

  o fix csv file detector

  o cxmag method: don't use set method which might have trouble with
    names which are numbers

  o add write_vcsv method: Cadence VCSV-format

  o revise/fix read_psf method

- DataViewx.py

  o add file->write VCSV-format

  o add file->rename columns

  o revise command parameter to include -x directive 

  o add y-target and x-target specifications for linear regression

  o add hex report of analog to digital parameters

- DialogBase.py

  o revise windows, resizable height only

- NGspice.py

  o add simulator configuration option (HSpice, NGspice)

  o file menu options: add write NGspice circuit file

  o edit menu: add simulator radiobuttons

  o fix xcol, ycols plotting 

- Pattern.py

  o revise/correct PRBS pattern generator

- PlotBase.py

  o add x- and y-spread to line-parameters report

  o add more information for vertical lines

- Report.py

  o add date and user methods

- SelectionDialog.py

  o add entry_emacs_bindings to entry windows

- Tckt.py

  o Fix spectre instance regular expression

  o Fix spectre sine source 

v1.0.4,  3-12-2017
------------------

- added new classes:

  o Itemizer

  o CanvasIdentify

- Data.py and Dataviewx.py

  o Add interpolation of other columns to periods

  o added read_touchstone method

  o added THD analysis

  o fix low-pass filter

  o add rescale/overlay operation

  o add advance/delay operation

  o dY/dX binary method -> deriv

- Tckt.py

  o Add case2ckey method

- StatusDialog.py

  o Fix update option

- MessageDialog.py

  o add write button/function

- PlotBase.py

  o Add/Change annotation bitmaps

  o Add CanvasIdentify curve identifier, and turn-on/turn-off in settings

- PlotBasem.py, DataViewm.py, XYplotm.py, and Plotterm.py

  o Plots using matplotlib

v1.0.5,  6-19-2017
------------------

- re-release, since doc directory and CHANGES.txt didn't get uploaded

v1.0.6,  5-28-2018
------------------

- matplotlib is default: added flyback removal

- Data.py

  o Add append_numpy_array method

  o Revise read_touchstone method

  o Revise moving-average filter method

  o add "no window" for FFT

  o add header_lines_to_skip argument to read_ssv method

- DataViewm.py

  o Add matplotlib flyback line removal

  o For resize binding, add pending to prevent multiple calls while resizing

  o Add "dots and narrow lines" to settings window

  o Add "S" and "s" key-bindings for toggling squares and dots

- Itemizer.py

  o Add __len__ method for number of experiments

v1.1.0,  1-01-2019
------------------

- add support for both Python 3 and Python 2.7

- clean up code

- added new class:

  o GetEntryDialog - replaces tkinter.simpledialog

- Data.py

  o add find_rows_where_equal method

  o add samples method

  o add read_vcsv method

  o jitter method: average frequency calculation - use average of steps

  o read_csv: set values to zero if not present

- DataViewm.py

  o minor fixes

- FrameNotebook.py

  o tabs set to fixed size, long names on balloon windows

- PlotBasem.py

  o minor changes

- SmithChart.py

  o fix extra parameter required for _plot_curves method

- Tckt.py

  o put back simulate_ngspice

  o op report changes for different versions of spectre

v1.1.1,  1-03-2019
------------------

- re-release, minor correction to setup.py and test_all

v1.1.2,  1-03-2019
------------------

- re-release, minor correction to setup.py and Tckt.py

v1.1.3,  1-03-2019
------------------

- re-release, doc correction to setup.py


decida/Pattern.py
################################################################################
# CLASS    : Pattern
# PURPOSE  : generate a bit-sequence pattern
# AUTHOR   : Richard Booth
# DATE     : Sat Nov  9 11:23:03 2013
# -----------------------------------------------------------------------------
# NOTES    :
#
# LICENSE  : (BSD-new)
#
# This software is provided subject to the following terms and conditions,
# which you should read carefully before using the software.  Using this
# software indicates your acceptance of these terms and conditions.  If you
# do not agree with these terms and conditions, do not use the software.
#
# Copyright (c) 2013 Richard Booth. All rights reserved.
#
# Redistribution and use in source or binary forms, with or without
# modifications, are permitted provided that the following conditions are met:
#     * Redistributions of source code must retain the above copyright
#       notice, this list of conditions and the following Disclaimer
#       in each human readable file as well as in the documentation and/or
#       other materials provided with the distribution.
#     * Redistributions in binary form must reproduce the above copyright
#       notice, this list of conditions and the following Disclaimer in the
#       documentation and/or other materials provided with the distribution.
#     * Neither the name of Richard Booth nor the names of contributors
#       (those who make changes to the software, documentation or other
#       materials) may be used to endorse or promote products derived from
#       this software without specific prior written permission.
#
# Disclaimer
#
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
# INCLUDING, BUT NOT LIMITED TO, INFRINGEMENT AND THE IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
# ANY USE, MODIFICATION OR DISTRIBUTION OF THIS SOFTWARE IS SOLELY AT THE
# USERS OWN RISK.  IN NO EVENT SHALL RICHARD BOOTH OR CONTRIBUTORS
# BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, INCLUDING,
# BUT NOT LIMITED TO, CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
# OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
# ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
# IN NO EVENT SHALL THE AUTHORS OR DISTRIBUTORS BE LIABLE TO ANY PARTY
# FOR DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES
# ARISING OUT OF THE USE OF THIS SOFTWARE, ITS DOCUMENTATION, OR ANY
# DERIVATIVES THEREOF, EVEN IF THE AUTHORS HAVE BEEN ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.
#
# THE AUTHORS AND DISTRIBUTORS SPECIFICALLY DISCLAIM ANY WARRANTIES,
# INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.  THIS SOFTWARE
# IS PROVIDED ON AN "AS IS" BASIS, AND THE AUTHORS AND DISTRIBUTORS HAVE
# NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR
# MODIFICATIONS.
################################################################################
from __future__ import print_function
from builtins import str
from builtins import range
import math
import random
from decida.ItclObjectx import ItclObjectx

class Pattern(ItclObjectx) :
    """
    **synopsis**:

        Pattern generator.

        *Pattern* generates one of various bit-sequences and outputs the
        sequence as either a string of 1's and 0's, a piece-wise linear set of
        time, voltage pairs, or a set of time, binary pairs.

        The *Tckt* class uses *Pattern* to generate piece-wise linear voltage
        specifications for procedural simulation scripts.

    **constructor arguments**:

        **\*\*kwargs** (dict)

              keyword=value specifications:
              configuration-options

    **configuration options**:

        **v0** (float, default=0.0)

              low value of the signal.

        **v1** (float, default=1.0)

              high value of the signal.

        **delay** (float, default=0.0)

              delay before first bit.

        **edge** (float, default=50ps)

              rise and fall times in s.

        **period** (float, default=1ns)

              bit period in s.

        **pre** (str, default="")

              specify preamble to bit-pattern (bit-sequence)

        **post** (str, default="")

              specify suffix to bit-pattern (bit-sequence)

        **start_at_first_bit** (bool, default=False)

              Normally a pwl sequence starts at the common-mode.
              If start_at_first_bit=True, start at first bit value.

        **format** (str, default="pwl")

              Format of return list:

                "binary": return pattern only

                "time-binary": return list of time binary pairs

                "pwl": return piecewise linear waveform

    **example** (from test_Pattern): ::

        from decida.Pattern import Pattern

        p = Pattern(delay=4e-9, start_at_first_bit=True, edge=0.1e-9)
        pwl = p.prbs(length=50)
        print("t v")
        for t,v in zip(pwl[0::2], pwl[1::2]) :
            print(t, v)

    **public methods**:

        * public methods from *ItclObjectx*

    """
    #%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    # Pattern main
    #%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    #==========================================================================
    # METHOD  : __init__
    # PURPOSE : constructor
    #==========================================================================
    def __init__(self, **kwargs) :
        ItclObjectx.__init__(self)
        #----------------------------------------------------------------------
        # private variables:
        # from www.xilinx.com/support/documentation/application_notes/xapp052.pdf"
        #----------------------------------------------------------------------
        self.__poly = (
            [],            [],    [0],           [1],           [2],           [2],           [4],           [5],
            [5,4,3],       [4],   [6],           [8],           [5,3,0],       [3,2,0],       [4,2,0],       [13],
            [14,12,3],     [13],  [10],          [5,1,0],       [16],          [18],          [20],          [17],
            [22,21,16],    [21],  [5,1,0],       [4,1,0],       [24],          [26],          [5,3,0],       [27],
            [21,1,0],      [19],  [26,1,0],      [32],          [24],          [4,3,2,1,0],   [5,4,0],       [34],
            [37,20,18],    [37],  [40,19,18],    [41,37,36],    [42,17,16],    [43,41,40],    [44,25,24],    [41],
            [46,20,19],    [39],  [48,23,22],    [49,35,34],    [48],          [51,37,36],    [52,17,16],    [30],
            [54,34,33],    [49],  [38],          [57,37,36],    [58],          [59,45,44],    [60,5,4],      [61],
            [62,60,59],    [46],  [64,56,55],    [65,57,56],    [58],          [66,41,39],    [68,54,53],    [64],
            [65,24,18],    [47],  [72,58,57],    [73,64,63],    [74,40,39],    [75,46,45],    [76,58,57],    [69],
            [78,42,41],    [76],  [78,46,43],    [81,37,36],    [70],          [83,57,56],    [84,73,72],    [73],
            [86,16,15],    [50],  [88,71,70],    [89,7,6],      [90,79,78],    [90],          [72],          [83],
            [93,48,46],    [90],  [86],          [96,53,51],    [62],          [99,94,93],    [100,35,34],   [93],
            [102,93,92],   [88],  [90],          [104,43,41],   [76],          [107,102,101], [108,97,96],   [100],
            [109,68,66],   [103], [112,32,31],   [113,100,99],  [114,45,44],   [114,98,96],   [84],          [110],
            [112,8,1],     [102], [120,62,61],   [120],         [86],          [123,17,16],   [124,89,88],   [125],
            [125,100,98],  [123], [126],         [129,83,82],   [102],         [131,81,80],   [76],          [123],
            [134,10,9],    [115], [136,130,129], [135,133,130], [110],         [139,109,108], [120],         [141,122,121],
            [142,74,73],   [92],  [144,86,85],   [145,109,108], [120],         [147,39,38],   [96],          [147],
            [150,86,85],   [151], [151,26,24],   [153,123,122], [154,40,39],   [155,130,129], [156,131,130], [127],
            [158,141,140], [142], [160,74,73],   [161,103,102], [162,150,149], [163,134,133], [164,127,126], [160],
            [165,152,150],
        )
        #----------------------------------------------------------------------
        # configuration options:
        #----------------------------------------------------------------------
        self._add_options({
            "v0"          : [0.0,    None],
            "v1"          : [1.0,    None],
            "delay"       : [0.0,    None],
            "edge"        : [50e-12, None],
            "period"      : [1e-9,   None],
            "pre"         : ["",     None],
            "post"        : ["",     None],
            "start_at_first_bit" : [False,  None],
            "format"      : ["pwl",  None],
        })
        #----------------------------------------------------------------------
        # keyword arguments are all configuration options
        #----------------------------------------------------------------------
        for key, value in list(kwargs.items()) :
            self[key] = value
    #%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    # Pattern configuration option callback methods
    #%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    #%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    # Pattern user commands
    #%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    #==========================================================================
    # METHOD  : pattern
    # PURPOSE : specify bit pattern directly
    #==========================================================================
    def pattern(self, pattern) :
        """ specify and generate a bit pattern directly.

        **arguments**:

            **pattern** (str)

                a string of "1"'s and "0"'s

        **results**:

            * The *Pattern* pattern is set to the bit pattern.

            * Returns pattern in format specified by the *format*
              configuration option.

        """
        return self.format(pattern)
    #==========================================================================
    # METHOD : clock
    # PURPOSE : generate clock bit-sequence
    # ARGUMENTS :
    #   % length
    #       number of bits to generate
    #   % startbit (optional)
    #       first bit in the sequence; default=0
    #   % bits_level (optional)
    #       number of bits in each clock level
    #==========================================================================
    def clock(self, length, startbit=0, bits_level=1) :
        """ generate clock pattern (0, 1, 0, 1, 0, ...).

        **arguments**:

            **length** (int)

                The number of bits in the sequence

            **startbit** (int, default=0)

                The starting bit of the bit sequence

            **bits_level** (int, default=1)

                Number of bits in each clock level.

        **results**:

            * The *Pattern* pattern is set to a clock pattern, with
              the start bit set to the specified value.

            * Returns pattern in format specified by the *format*
              configuration option.

        """
        if bits_level < 1:
            bits_level = 1
        pattern = []
        bit = startbit
        while len(pattern) < length :
            bit = 1-bit
            for j in range(0, bits_level) :
                pattern.append(str(bit))
        pattern=pattern[0:length]
        return self.format("".join(pattern))
    #==========================================================================
    # METHOD : long_short
    # PURPOSE : generate long 1/0 followed by clock
    #==========================================================================
    def long_short(self, longs, shorts, startbit=0, length=0) :
        """ generate a bit pattern with several repeated bits, followed by
            a phase where each bit is inverted.

        **arguments**:

            **longs** (int)

                number of bits in the long phase, where each bit
                is repeated.

            **shorts** (int)

                number of bits in the short phase, where each bit
                is inverted.

            **startbit** (int, default=0)

                The starting bit of the bit sequence

            **length** (int, default=0)

                The number of bits in the sequence.
                If not specified, or less than 1, set to long+short

        **results**:

            * The *Pattern* pattern is set to the bit pattern.

            * Returns pattern in format specified by the *format*
              configuration option.

        """
        if length < 1:
            length = longs+shorts
        pattern = []
        bit = startbit
        while len(pattern) < length :
            for j in range(0, longs) :
                pattern.append(str(bit))
            for j in range(0, shorts) :
                bit = 1 - bit
                pattern.append(str(bit))
        pattern = pattern[0:length]
        return self.format("".join(pattern))
    #==========================================================================
    # METHOD : rand
    # PURPOSE : generate random bit-sequence
    # ARGUMENTS :
    #   % length
    #       number of bits to generate
    #   % seed
    #       random seed (default = None)
    #==========================================================================
    def rand(self, length, seed=None) :
        """ generate a random bit sequence.

        **arguments**:

            **length** (int)

                the length of the bit sequence.

            **seed** (str, default=None)

                random seed to initialize random number generator.

        **results**:

            * The *Pattern* pattern is set to a random bit pattern.

            * Returns pattern in format specified by the *format*
              configuration option.

        """
        if seed :
            random.seed(seed)
        pattern = []
        for i in range(0, length) :
            bit = 1 if random.random() > 0.5 else 0
            pattern.append(str(bit))
        return self.format("".join(pattern))
    #==========================================================================
    # METHOD : prbs
    # ARGUMENTS :
    #==========================================================================
    def prbs(self, size=7, length=0) :
        """ generate a pseudo-random bit pattern.

        **arguments**:

            **size** (int, default=7)

                The PRBS specification: size=7 generates a pseudo-random
                bit-sequence of length 2^7-1.

            **length** (int, default=0)

                The length of the bit-sequence.  If length=0, generates
                a sequence of 2^size - 1.

        **results**:

            * The *Pattern* pattern is set to the PRBS bit pattern.

            * Returns pattern in format specified by the *format*
              configuration option.

        """
        if size < 2 or size > 168:
            self.error("size must be between 2 and 168")
            return None
        taps = [int(x) for x in self.__poly[size]]
        if length < 1 :
            length = int(math.pow(2, size)) - 1
        register = [0] * size
        pattern  = []
        for i in range(0, length) :
            acc = register[-1] ^ 1
            pattern.append(str(acc))
            for tap in taps :
                acc = register[tap] ^ acc
            register.pop(-1)
            register.insert(0, acc)
        return self.format("".join(pattern))
    #==========================================================================
    # METHOD : format
    # ARGUMENTS :
    #   % pattern
    #       generated bit sequence
    #==========================================================================
    def format(self, pattern) :
        """ format the bit-sequence.

        **arguments**:

            **pattern** (string)

                a string of "1"'s and "0"'s

        **results**:

            * Returns *Pattern* pattern in format specified by the *format*
              configuration option.

              format is one of:

                * "binary": string sequence of "1"'s and "0"'s

                * "time-binary": pairs of time, and integer value 1 or 0.

                * "pwl": pairs of time, value to account for signal
                  edge, period, and voltage low and high values.

        """
        pat = self["pre"] + pattern + self["post"]
        if   self["format"] == "binary" :
            return pat
        elif self["format"] == "time-binary" :
            time = 0.0
            olist = []
            for sbit in pat :
                olist.extend((time, int(sbit)))
                time += self["period"]
            return olist
        elif self["format"] == "pwl" :
            start = True
            time  = 0.0
            olist = []
            vcm   = (self["v1"] + self["v0"]) * 0.5
            for sbit in pat :
                vbit = self["v1"] if sbit == "1" else self["v0"]
                vbit_1 = vbit
                if start :
                    start = False
                    if self["start_at_first_bit"] :
                        if self["delay"] > 0.0:
                            olist.extend((time, vbit))
                            time += self["delay"]
                        olist.extend((time, vbit))
                    else :
                        if self["delay"] > 0.0:
                            olist.extend((time, vcm))
                            time += self["delay"]
                        olist.extend((time, vcm))
                        time += self["edge"]*0.5
                        olist.extend((time, vbit))
                else :
                    time += self["period"] - self["edge"]
                    olist.extend((time, vbit_1))
                    time += self["edge"]
                    olist.extend((time, vbit))
                vbit_1 = vbit
            return olist
        return None
